### ISSUE TITLE
AIs stop recruiting attacking units after the defense units reach the maximum size!

### ISSUE THREAD

**Issue by Admiral-Gumpert** [Sh0wdown/UnofficialCrusaderPatch#599]
I have observed a strange bug which could also be related to this issue #104 
It might also explain why this issue even exists , If an AI keeps recruiting its defense units until they reach the max he will stop raiding or attacking , (because defense units cannot be recruited anymore)
![image](https://user-images.githubusercontent.com/50601775/79897815-ee450580-8412-11ea-8495-74122f2efc8f.png)

Here you can see wolf isn't recruiting any units , even you can see workers cannot go to the armoury because its already full.
I dont remember facing this issue on any previous versions ( I used to have probabilities for defending and attacking to sum up over 100, but now I made them back to 100 because it won't work anymore).

**Comment by Truetobi91** [comment]
Are you playing on extreme?

**Comment by Admiral-Gumpert** [comment]
Yes, you can see the magic bar on the right.

**Comment by J-T-de** [comment]
I don't think this is the complete story, as there are cases when the defense is fully equipped and the attacks continue. Is this reproducible with your map/ai? Does saving and loading fix the problem?

**Comment by Krarilotus** [comment]
The trigger for this needs to be tested. If you want to test it for us, feel free to join our tester squad (we need more testers!!!) on our discord :D

**Comment by Admiral-Gumpert** [comment]
> I don't think this is the complete story, as there are cases when the defense is fully equipped and the attacks continue. Is this reproducible with your map/ai? Does saving and loading fix the problem?

yeah saving and loading the map doesnt fix this unfortunately, I'm pretty sure it is caused by maxing defense units, I have tested it 3 times so far.

**Comment by Truetobi91** [comment]
Could it be possible that the AI can only handle a certain amount of defensive units, breaking its behaviour when exceeding it? Or in other words, the parameter doesn't have an infinite number of valid values?

There is - imo - a comparable thing with the maxstone parameters just as one example. If the value is below 64 (not 100% sure but I think so) the AI will store as much stone as set by this parameter. If you set the parameter higher than that, the AI will store more stone than the parameter said. E.g I tested values of 70 oder 80 and the AI would then store like 90 and 110 (I don't remember the exact numbers because it's been a while since I noticed that behavior).

**Comment by Admiral-Gumpert** [comment]
Ok here's the thing, Maps which was created in normal stronghold aren't the same as the ones created in the extreme version, Hexadecimal data are different and also unit caps are also different (not just units but also any kind of animated object),

**Comment by Krarilotus** [comment]
If you start a game on a normal map in extreme, the file format will automatically get changed and adapts the bigger army size.

If you want to trace the bug, i recommend trying out my first AI contest, where richard reproduceable always stops recruitimg after reaching 400 troops.

For the stone and iron in the stockpile, that gets stored extra: the amount depends on the most expensive building the AI has in its castle, hence if he has a boilimg pot, he might store 10 extra iron. If he has a big round tower, he might store 40 extra stone etc.

**Comment by Truetobi91** [comment]
Getting back to the bug with an explanation for that bevahoir:

If the AI's army includes a unit outside their main army (e.g. MaxTunnelers, MaxAssassins; AttUnit2 etc) with a number >0 that they cannot recruit this bug will appear at a very specific and reproducable point which is caused by how the "soft cap" (as I call it) of AttMaxDefault works.

The AI recruits their mix normal until reaching that cap. If they cannot recruit a unit they will skip it. Once the main army has reached their soft cap the AI is forced to only recruit the units of the other attacking roles (ALL of them). If they cannot recruit a certain unit from these slots, e.g. having tunnelers set to 10 when the AI doesn't even have a tunnelers guild it causes the AI to get stuck. 
They cannot recruit anything else because the other attacking slots are maxed out but the soft cap still prevents the AI from recruiting any more main army until these tunnelers are recruitet. And since the AI doesn't reach their next attack wave size it won't attack anymore.

**Comment by Krarilotus** [comment]
@Truetobi thanks for that explanation ,that makes a lot of sense! Is it
also possible that an AI would be able to recruit a unit, but doesn't? I
feel like that could happen with engineers for example. I would like to
have some testing on how engineers behave when attackwaves can't reach
their size because of them being in the way for recruitment, even tho the
maximum amount of siege equipment hasn't been built yet. And also do the
engineers strictly count into the army size for attacking or are they
counted extra?
